Redesign preparePinnedVirtualThreadForUnmount logic

fengxue-IS opened the pull request:
Remove use of objectMonitorInflate on lock that current thread doesn't own.
Change to use of FLC bit to signal owner thread to inflate the monitor
when it exits the monitor.

Use CAS operation to ensure only one thread can update the lockword flag bits
to contended flatlock and ignore current lock learning state.

Lock reservation is cancelled before inflation/FLC bit set.

keithc-ca commented on runtime/codert_vm/cnathelp.cpp:
```diff
@@ -1716,6 +1716,11 @@ slow_jitMonitorEnterImpl(J9VMThread *currentThread, bool forMethod)
 		/* Try to yield the virtual thread if it will be blocked. */
 		monstatus = currentThread->javaVM->internalVMFunctions->preparePinnedVirtualThreadForUnmount(currentThread, (j9object_t)currentThread->floatTemp2, false);
+		if (monstatus > J9_OBJECT_MONITOR_BLOCKING) {
+			/* Monitor have been acquired. */
```
have -> has

keithc-ca commented on runtime/oti/vm_api.h:
```diff
@@ -4734,7 +4734,8 @@ preparePinnedVirtualThreadForMount(J9VMThread *currentThread, j9object_t continu
  * @param isObjectWait if the call is from Object.wait()
  *
- * @return J9_OBJECT_MONITOR_YIELD_VIRTUAL if the can be successfully yielded;
+ * @return syncObj if isObjectWait is false and monitor can acquired;
+ * J9_OBJECT_MONITOR_YIELD_VIRTUAL if the can be successfully yielded;
```
monitor can acquired -> monitor can be acquired
Perhaps "monitor" is missing in "if the can be"?

keithc-ca commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -1539,6 +1539,11 @@ class INTERPRETER_CLASS
 			omrthread_monitor_enter(_vm->blockedVirtualThreadsMutex);
 			_currentThread->currentContinuation->nextWaitingContinuation = _vm->blockedContinuations;
 			_vm->blockedContinuations = _currentThread->currentContinuation;
+
+			if (NULL == _currentThread->currentContinuation->objectWaitMonitor->monitor->owner) {
+				/* notify unblocker if the blocking monitor is unlocked. */
+				omrthread_monitor_notify(_vm->blockedVirtualThreadsMutex);
+			}
```
Perhaps a local for `_currentThread->currentContinuation` (it's used three times here)?

keithc-ca commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
 	J9JavaVM *vm = currentThread->javaVM;
 
 	if (NULL != syncObj) {
+		j9objectmonitor_t volatile *lwEA = VM_ObjectMonitor::inlineGetLockAddress(currentThread, syncObj);;
```
double semi-colon

keithc-ca commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
+			} else {
+				/* This must be a monitor enter case, so this implies that a monitor entry was
+				 * created as the non-blocking path would have failed.
+				 * Since a monitor can only be inflated by a thread that owns it, we can't direct
```
direct -> directly

keithc-ca commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
+				/* Try acquire the inflated monitor */
+				if (0 == omrthread_monitor_try_enter(monitor)) {
+					/* if the INFLATED bit is set, then it is already inflated and we own the object monitor. */
+					if (J9_ARE_ANY_BITS_SET(((J9ThreadMonitor*)monitor)->flags, J9THREAD_MONITOR_INFLATED)) {
```
Space before `*`, please.

keithc-ca commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
+						j9objectmonitor_t newLockword = 0;
+						UDATA count = J9_FLATLOCK_COUNT(lock);
+						if (0 == count) {
+							/* lock is unlocked, so try to directly acquire it as a flatlock. */
```
Please start each sentence with a capital; see also lines 988, 1000, 1008 and 1016.

keithc-ca commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
+						} else {
+							/* change the lock to flat with FLC bit set so it will be inflated during exit. */
+							newLockword = (J9_FLATLOCK_OWNER(lock)
+										| (count - 1) << OBJECT_HEADER_LOCK_V2_RECURSION_OFFSET
```
Missing parentheses.

keithc-ca commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
+		j9objectmonitor_t volatile *lwEA = VM_ObjectMonitor::inlineGetLockAddress(currentThread, syncObj);
+		j9objectmonitor_t lock = J9_LOAD_LOCKWORD(currentThread, lwEA);
+		omrthread_monitor_t monitor = NULL;
+		
```
Please address the line-end check failure.

keithc-ca commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
+							newLockword = (j9objectmonitor_t)(UDATA)currentThread;
+						} else {
+							/* Change the lock to flat with FLC bit set so it will be inflated during exit. */
+							newLockword = ((j9objectmonitor_t)J9_FLATLOCK_OWNER(lock)
+										| ((j9objectmonitor_t)(count - 1) << OBJECT_HEADER_LOCK_V2_RECURSION_OFFSET)
+										| OBJECT_HEADER_LOCK_FLC);
```
The type of `newLockword` should be `UDATA`; the cast to `j9objectmonitor_t` should be done on line 994 (`|` is not well-defined for pointers).

JasonFengJ9 commented:
This PR makes progress on [jdk24 java/lang/Thread/virtual/RetryMonitorEnterWhenPinned.java timeout](https://github.com/eclipse-openj9/openj9/issues/20955), now the test hang at [wait for all threads to terminate](https://github.com/ibmruntimes/openj9-openjdk-jdk24/blob/a9802a89c663d70223765d7af8c41e83db582199/test/jdk/java/lang/Thread/virtual/RetryMonitorEnterWhenPinned.java#L107-L108)

hangshao0 commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
 	J9JavaVM *vm = currentThread->javaVM;
 
 	if (NULL != syncObj) {
+		j9objectmonitor_t volatile *lwEA = VM_ObjectMonitor::inlineGetLockAddress(currentThread, syncObj);
```
Should hooks like `J9HOOK_VM_MONITOR_CONTENDED_ENTER` be triggered somewhere in this function ?

fengxue-IS replied:
Will create a separate PR to fix the jvmti contended enter hook. That will require some additional changes to store the monitor state when yielding.

fengxue-IS commented:
Addressed review comments, opened a new issue on creating common helper (will be pursued after JDK24 release). FYI @tajila 

@keithc-ca  can you please take another look, this PR is now ready for review.

keithc-ca commented on runtime/vm/ContinuationHelpers.cpp:
```diff
@@ -923,14 +923,110 @@ preparePinnedVirtualThreadForUnmount(J9VMThread *currentThread, j9object_t syncO
+					/* Inflated monitor owned by another thread, so the lockword update will be completed by them. */
+					lock = J9_LOAD_LOCKWORD(currentThread, lwEA);
+					/* We can safely continue if the lock is inflated or FLC bit is set. */
+					if (!J9_LOCK_IS_INFLATED(lock)
+					&& (OBJECT_HEADER_LOCK_FLC != (lock & (OBJECT_HEADER_LOCK_FLC | OBJECT_HEADER_LOCK_INFLATED)))
+					) {
```
This can be simplified: `!J9_LOCK_IS_INFLATED(lock)` implies `OBJECT_HEADER_LOCK_INFLATED` is not set, in which case the second condition simplifies to `OBJECT_HEADER_LOCK_FLC != (lock & OBJECT_HEADER_LOCK_FLC)`. The two
conditions are true if and only if `0 == (lock & (OBJECT_HEADER_LOCK_FLC | OBJECT_HEADER_LOCK_INFLATED))` (which more closely matches the comment).
```c
					if (0 == (lock & (OBJECT_HEADER_LOCK_FLC | OBJECT_HEADER_LOCK_INFLATED))) {
```

keithc-ca commented:
Could you please put more information in the commit message (and the description here) that explains (even if only at a very high level) what is different in the new design?

keithc-ca commented:
Jenkins test sanity,extended plinux jdk24
